File: ios/SecretaryVaultAccess/Sources/SecretaryVaultAccess/SyncModels.swift
import Foundation

/// One device's vector-clock entry — public metadata, never secret.
/// Mirrors the bridge `DeviceClockDto`.
public struct DeviceClock: Equatable {
    public let deviceUuidHex: String
    public let counter: UInt64
    public init(deviceUuidHex: String, counter: UInt64) {
        self.deviceUuidHex = deviceUuidHex
        self.counter = counter
    }
}

/// Read-only sync status for a vault. Mirrors the bridge `SyncStatusDto`.
public struct SyncStatus: Equatable {
    public let hasState: Bool
    public let deviceClocks: [DeviceClock]
    public let lastStateWriteMs: UInt64?
    public init(hasState: Bool, deviceClocks: [DeviceClock], lastStateWriteMs: UInt64?) {
        self.hasState = hasState
        self.deviceClocks = deviceClocks
        self.lastStateWriteMs = lastStateWriteMs
    }
}

/// A tombstone dispute awaiting a human decision. Metadata-only by construction
/// (the bridge projects field *names*, never values) — no plaintext secret here.
public struct SyncVeto: Equatable {
    public let recordUuidHex: String
    public let recordType: String
    public let tags: [String]
    public let fieldNames: [String]
    public let localLastModMs: UInt64
    public let peerTombstonedAtMs: UInt64
    public let peerDeviceHex: String
    public init(recordUuidHex: String, recordType: String, tags: [String],
                fieldNames: [String], localLastModMs: UInt64,
                peerTombstonedAtMs: UInt64, peerDeviceHex: String) {
        self.recordUuidHex = recordUuidHex
        self.recordType = recordType
        self.tags = tags
        self.fieldNames = fieldNames
        self.localLastModMs = localLastModMs
        self.peerTombstonedAtMs = peerTombstonedAtMs
        self.peerDeviceHex = peerDeviceHex
    }
}

/// Metadata-only field-level collision summary for the "auto-merged" notice.
public struct SyncCollision: Equatable {
    public let recordUuidHex: String
    public let fieldNames: [String]
    public init(recordUuidHex: String, fieldNames: [String]) {
        self.recordUuidHex = recordUuidHex
        self.fieldNames = fieldNames
    }
}

/// Caller's per-record decision. `keepLocal == true` rejects the peer tombstone;
/// `false` accepts the delete.
public struct SyncVetoDecision: Equatable {
    public let recordUuidHex: String
    public let keepLocal: Bool
    public init(recordUuidHex: String, keepLocal: Bool) {
        self.recordUuidHex = recordUuidHex
        self.keepLocal = keepLocal
    }
}

/// Result of one sync pass. Mirrors the bridge `SyncOutcomeDto`.
public enum SyncOutcome: Equatable {
    /// Disk clock == local highest-seen. No change.
    case nothingToDo
    /// Disk strictly dominates local. State advanced; no vault write.
    case appliedAutomatically
    /// Concurrent but no surviving block divergence. State advanced; no write.
    case silentMerge
    /// Concurrent, diverging, zero vetoes → merged result committed.
    case mergedClean
    /// Disk strictly dominated by local (rollback). Nothing changed.
    case rollbackRejected
    /// Concurrent, diverging, tombstone vetoes need a human. NOT committed.
    /// `manifestHash` is the opaque freshness token to pass back to commit.
    case conflictsPending(vetoes: [SyncVeto], collisions: [SyncCollision], manifestHash: [UInt8])
}

/// The pending conflict detail surfaced after a `runPass` that paused.
/// The freshness token is held privately by the coordinator and intentionally
/// NOT exposed here — callers never thread it themselves.
public struct PendingConflict: Equatable {
    public let vetoes: [SyncVeto]
    public let collisions: [SyncCollision]
    public init(vetoes: [SyncVeto], collisions: [SyncCollision]) {
        self.vetoes = vetoes
        self.collisions = collisions
    }
}

File: android/vault-access/src/main/kotlin/org/secretary/sync/SyncModels.kt
package org.secretary.sync

/** Vector-clock entry. Never secret — public metadata. */
data class DeviceClock(val deviceUuidHex: String, val counter: ULong)

/** Read-only sync status snapshot. */
data class SyncStatus(
    val hasState: Boolean,
    val deviceClocks: List<DeviceClock>,
    val lastStateWriteMs: ULong?,
)

/**
 * Tombstone-dispute projection. Field [fieldNames] are NAMES only — never values
 * (anti-oracle / metadata-only discipline).
 */
data class SyncVeto(
    val recordUuidHex: String,
    val recordType: String,
    val tags: List<String>,
    val fieldNames: List<String>,
    val localLastModMs: ULong,
    val peerTombstonedAtMs: ULong,
    val peerDeviceHex: String,
)

/** Field-level last-writer-wins collision notice (field NAMES only). */
data class SyncCollision(val recordUuidHex: String, val fieldNames: List<String>)

/** The caller's per-record veto decision. `keepLocal == true` rejects the peer tombstone. */
data class SyncVetoDecision(val recordUuidHex: String, val keepLocal: Boolean)

/** A paused pass's conflict detail, surfaced for interactive resolution. */
data class PendingConflict(val vetoes: List<SyncVeto>, val collisions: List<SyncCollision>)

/**
 * Result of one sync pass. Arms map 1:1 to the uniffi `SyncOutcomeDto` so the future
 * `UniffiVaultSyncPort` adapter is a straight transcription.
 */
sealed interface SyncOutcome {
    data object NothingToDo : SyncOutcome
    data object AppliedAutomatically : SyncOutcome
    data object SilentMerge : SyncOutcome
    data object MergedClean : SyncOutcome
    data object RollbackRejected : SyncOutcome

    /**
     * A tombstone dispute paused the pass without writing. [manifestHash] is the opaque
     * TOCTOU freshness token replayed into `commitDecisions`. Not a `data class`: it carries
     * a [ByteArray], so equality/hashing are content-based via explicit overrides.
     */
    class ConflictsPending(
        val vetoes: List<SyncVeto>,
        val collisions: List<SyncCollision>,
        val manifestHash: ByteArray,
    ) : SyncOutcome {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is ConflictsPending) return false
            return vetoes == other.vetoes &&
                collisions == other.collisions &&
                manifestHash.contentEquals(other.manifestHash)
        }

        override fun hashCode(): Int {
            var result = vetoes.hashCode()
            result = 31 * result + collisions.hashCode()
            result = 31 * result + manifestHash.contentHashCode()
            return result
        }

        override fun toString(): String =
            "ConflictsPending(vetoes=$vetoes, collisions=$collisions, " +
                "manifestHash=${manifestHash.size} bytes)"
    }
}
